### PR TITLE
feat(kzip): add subcommand and proto for metadata kzips

### PR DESCRIPTION
**Pull request by justbuchanan**
This lets us store the commit timestamp of the build in a CompilationUnit:

```json
{
  "v_name": {
    "language": "kythe_build_metadata"
  },
  "details": [
    {
      "@type": "kythe.io/proto/kythe.proto.BuildMetadata",
      "commit_timestamp": "2020-06-05T06:15:09Z"
    }
  ]
}
```

I put the new `BuildMetadata` message in `storage.proto`, but maybe there's a better place?

### CODE REVIEW

File: kythe/proto/storage.proto [hunk 1]
```diff
@@ -203,3 +204,10 @@ message VNameRewriteRule {
 message VNameRewriteRules {
   repeated VNameRewriteRule rule = 1;
 }
+
+// RepoMetadata provides storage for additional information about a specific
+// build from a repository. This message is stored in the
+// CompilationUnit.details field.
+message BuildMetadata {
```
Comment (jaysachs) [inline kythe/proto/storage.proto:211]: Maybe this belongs in analysis.proto alongside CompilationUnit? Or maybe buildinfo.proto?

File: kythe/proto/storage.proto [hunk 2]
```diff
@@ -203,3 +204,10 @@ message VNameRewriteRule {
...
+  google.protobuf.Timestamp commit_timestamp = 1;
```
Comment (jaysachs) [inline kythe/proto/storage.proto:212]: Is it clear what this is a "commit" of? What if the build comes from multiple repositories? Should we be more generic, and use something like "origin_timestamp"?

Comment (justbuchanan) [reply kythe/proto/storage.proto:212]: `origin_timestamp` sounds fine to me. Or we can clarify that it's the commit timestamp of the "superproject"?

File: kythe/proto/storage.proto (hunk 1 above)
Comment (justbuchanan) [reply kythe/proto/storage.proto]: Moved to analysis.proto.

File: kythe/go/platform/tools/kzip/metadatacmd/metadatacmd.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+		return c.Fail("Creating output file: %v", err)
+	}
+	opt := kzip.WithEncoding(c.encoding.Encoding)
+	w, err := kzip.NewWriteCloser(f, opt)
+	if err != nil {
+		f.Close()
+		return c.Fail("Failed to close writer: %v", err)
+	}
+	if _, err := w.AddUnit(unit, nil); err != nil {
+		return c.Fail("Failed to add unit", err)
```
Comment (jaysachs) [inline kythe/go/platform/tools/kzip/metadatacmd/metadatacmd.go:94]: w.close() here?

File: kythe/proto/analysis.proto [hunk 4]
```diff
@@ -211,6 +212,13 @@ message KzipInfo {
   repeated string critical_kzip_errors = 5;
 }
 
+// RepoMetadata provides storage for additional information about a specific
+// build from a repository. This message is stored in the
+// CompilationUnit.details field.
+message BuildMetadata {
+  google.protobuf.Timestamp commit_timestamp = 1;
```
Comment (jaysachs) [inline kythe/proto/analysis.proto:219]: Let's document that this is reasonably used for the "primary repository" commit, but could just as easily be some other origin timestamp.

File: kythe/go/platform/kzip/buildmetadata/buildmetadata.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	meta := &apb.BuildMetadata{
+		CommitTimestamp: tp,
+	}
+	det, err := kptypes.MarshalAny(meta)
+	if err != nil {
+		return nil, fmt.Errorf("marshaling details: %v", err)
+	}
+	return &apb.CompilationUnit{
+		VName: &spb.VName{
+			Language: Language,
```
Comment (jaysachs) [inline kythe/go/platform/kzip/buildmetadata/buildmetadata.go:54]: Hmm .. should this have a corpus? Or optionally have a corpus?

File: kythe/go/platform/tools/kzip/metadatacmd/metadatacmd.go (hunk 3 above)
Comment (justbuchanan) [reply kythe/go/platform/tools/kzip/metadatacmd/metadatacmd.go:94]: :+1:

File: kythe/go/platform/kzip/buildmetadata/buildmetadata.go (hunk 5 above)
Comment (justbuchanan) [reply kythe/go/platform/kzip/buildmetadata/buildmetadata.go:54]: Done.

File: kythe/go/platform/tools/kzip/metadatacmd/metadatacmd.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	w, err := kzip.NewWriteCloser(f, opt)
+	if err != nil {
+		f.Close()
+		return c.Fail("Failed to close writer: %v", err)
+	}
+	defer w.Close()
+	if _, err := w.AddUnit(unit, nil); err != nil {
+		return c.Fail("Failed to add unit", err)
+	}
+	if err := w.Close(); err != nil {
```
Comment (jaysachs) [inline kythe/go/platform/tools/kzip/metadatacmd/metadatacmd.go:105]: It's weird having the `defer w.Close()` and this explicit invocation.

File: kythe/go/platform/tools/kzip/metadatacmd/metadatacmd.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+	}
+	if c.commitTimestamp == "" {
+		return c.Fail("Required --commit_timestamp missing")
+	}
+
+	t, err := time.Parse(c.timestampFormat, c.commitTimestamp)
+	if err != nil {
+		return c.Fail("Unable to parse timestamp: '%s'", c.commitTimestamp)
+	}
+	unit, err := buildmetadata.CreateMetadataUnit(t)
```
Comment (jaysachs) [inline kythe/go/platform/tools/kzip/metadatacmd/metadatacmd.go:81]: Seems like corpus should maybe be supplied to `CreateMetadataUnit`?

File: kythe/go/platform/tools/kzip/metadatacmd/metadatacmd.go (hunk 6 above)
Comment (justbuchanan) [reply kythe/go/platform/tools/kzip/metadatacmd/metadatacmd.go:103]: Agreed. Changed to just a `w.Close()` in the if statement. I could do a defer and set the return value by name, but it's kinda clunky.

File: kythe/go/platform/tools/kzip/metadatacmd/metadatacmd.go (hunk 7 above)
Comment (justbuchanan) [reply kythe/go/platform/tools/kzip/metadatacmd/metadatacmd.go]: Done.

**Comment by justbuchanan** [comment]
I'm not sure why the build is failing - it works on my machine...

Here are the logs before it dies:

```shell
link: warning: package "github.com/golang/protobuf/ptypes/any" is provided by more than one rule:
    @com_github_golang_protobuf//ptypes/any:go_default_library
    @io_bazel_rules_go//proto/wkt:any_go_proto
Set "importmap" to different paths in each library.
This will be an error in the future.
link: warning: package "github.com/golang/protobuf/ptypes/duration" is provided by more than one rule:
    @com_github_golang_protobuf//ptypes/duration:go_default_library
    @io_bazel_rules_go//proto/wkt:duration_go_proto
Set "importmap" to different paths in each library.
This will be an error in the future.
link: warning: package "github.com/golang/protobuf/ptypes/timestamp" is provided by more than one rule:
    @com_github_golang_protobuf//ptypes/timestamp:go_default_library
    @io_bazel_rules_go//proto/wkt:timestamp_go_proto
Set "importmap" to different paths in each library.
This will be an error in the future.
link: warning: package "github.com/golang/protobuf/ptypes" is provided by more than one rule:
    @com_github_golang_protobuf//ptypes:go_default_library_gen
    @com_github_golang_protobuf//ptypes:go_default_library
Set "importmap" to different paths in each library.
This will be an error in the future.
panic: runtime error: index out of range [23] with length 0
goroutine 1 [running]:
cmd/link/internal/ld.decodetypeKind(...)
	/usr/local/go/src/cmd/link/internal/ld/decodesym.go:69
cmd/link/internal/ld.newtype(0xc000630180, 0xc002511010, 0xc0034c7800)
	/usr/local/go/src/cmd/link/internal/ld/dwarf.go:425 +0x2a32
cmd/link/internal/ld.defgotype(0xc000630180, 0xc002511010, 0x36)
	/usr/local/go/src/cmd/link/internal/ld/dwarf.go:420 +0x1c8
cmd/link/internal/ld.dwarfGenerateDebugInfo(0xc000630180)
	/usr/local/go/src/cmd/link/internal/ld/dwarf.go:1756 +0xa22
cmd/link/internal/ld.Main(0x89a320, 0x10, 0x20, 0x1, 0x7, 0x10, 0x6c35da, 0x1b, 0x6bf65d, 0x14, ...)
	/usr/local/go/src/cmd/link/internal/ld/main.go:219 +0xb96
main.main()
	/usr/local/go/src/cmd/link/main.go:68 +0x1bc
link: error running subcommand: exit status 2
```

The warning about proto packages being provided more than once is unfortunate, but I don't *think* this PR makes that worse than it already is.

**Pull request merged**